### PR TITLE
[WIP] Runtime modules patch

### PR DESCRIPTION
**Pull request by Teemperor**
Just for people to see the code. Lots of hack and debug code, so no comments like "Don't include <iostream>!!!!" please

### CODE REVIEW

File: core/dictgen/src/rootcling_impl.cxx [hunk 1]
```diff
@@ -4225,9 +4309,27 @@ int RootClingMain(int argc,
 
    ROOT::TMetaUtils::SetPathsForRelocatability(clingArgs);
 
+   bool isPCH = (sharedLibraryPathName == "allDict.cxx");
+   auto clingArgsModules = clingArgs;
+   if (!isPCH && getenv("ROOT_MODULES")) {
+     clingArgsModules.push_back("-fmodules");
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: We will probably need -fcxx-modules as well if we are on OSX.

File: core/dictgen/src/rootcling_impl.cxx [hunk 2]
```diff
@@ -4316,6 +4419,11 @@ int RootClingMain(int argc,
       }
    }
 
+   interp.declare("typedef double         Double32_t;");
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: Did you mean `#include "RtypesCore.h"`?

File: core/metacling/src/TCling.cxx [hunk 3]
```diff
@@ -1100,6 +1100,10 @@ TCling::TCling(const char *name, const char *title)
            eArg = clingArgsStorage.end(); iArg != eArg; ++iArg)
       interpArgs.push_back(iArg->c_str());
 
+   if (!fromRootCling && getenv("ROOT_MODULES")) {
+     interpArgs.push_back("-fmodules");
```
Comment (vgvassilev) [inline core/metacling/src/TCling.cxx]: Likewise, we need `-fcxx-modules` for OSX.

File: interpreter/cling/lib/Interpreter/CIFactory.cpp [hunk 4]
```diff
@@ -728,6 +763,33 @@ static void stringifyPreprocSetting(PreprocessorOptions& PPOpts,
     std::vector<const char*> argvCompile(argv, argv+1);
     argvCompile.reserve(argc+5);
 
+    // Just for temporal storage
+    std::string overlayArg;
+    std::string cacheArg;
+    std::string moduleNameArg;
+
+    if (COpts.CxxModules && !OnlyLex) {
+      argvCompile.push_back("-fmodules");
```
Comment (vgvassilev) [inline interpreter/cling/lib/Interpreter/CIFactory.cpp]: We will probably need `-fcxx-modules` as well if we are on OSX.

File: core/dictgen/src/rootcling_impl.cxx [hunk 5]
```diff
@@ -2238,6 +2242,42 @@ static bool GenerateAllDict(TModuleGenerator &modGen, clang::CompilerInstance *c
...
+      // will have a nicer way to do this in the future.
+      // NOTE: This is on purpose '4', not '5' which is the size of the
+      // vector. The last element is the list of excluded headers which we
+      // obviously don't want to check here.
+      for (int i = 0; i < 4; i++) {
+         auto &headerList = module->Headers[i];
+         for (const clang::Module::Header &moduleHeader : headerList) {
+            auto result = interpreter.declare("#include \"" + moduleHeader.NameAsWritten + "\"\n");
+            if (result != cling::Interpreter::CompilationResult::kSuccess) {
+              llvm::errs() << "Failed to include\n";
```
Comment (vgvassilev) [inline core/dictgen/src/rootcling_impl.cxx]: Why not converting the if-stmt into an assert?

**Comment by vgvassilev** [comment]
We should probably move 'Revert cling patch that broke using ParentClass::Method;' in a separate PR to test if it breaks anything in regular builds.

**Comment by martinmine** [comment]
@phsft-bot build

File: core/dictgen/res/Scanner.h [hunk 6]
```diff
@@ -46,6 +46,8 @@ class ClassSelectionRule;
 class RScanner: public clang::RecursiveASTVisitor<RScanner>
 {
 
+  bool canVisitDecl(clang::NamedDecl *D);
```
Comment (vgvassilev) [inline core/dictgen/res/Scanner.h]: We use the notion of `shouldVisit*` in many other places.

**Comment by vgvassilev** [comment]
@Teemperor, could you rebase?

**Comment by Teemperor** [comment]
We are already ahead of this branch in master + open PRs, so no more reason to keep this around.

**Pull request closed without merging**